### PR TITLE
optimize select query from cluster table function

### PR DESCRIPTION
**Pull request by liangliangpan**
Use local node as first priority to get Structure Of Remote Table.
we have many distributed queries( like   select xx from cluster('xx',view  (xxxx)   ) on a clickhouse cluster.  we found that the first node (shard_num=1)  have 2 times of query number compared to other shards.
The reason is that  the getStructureOfRemoteTableInShard func always take the first shard to execute  "DESC TABLE xx" query.
The better way is to  use local node as first priority which save the network rpc and reduce the pressure of first shard .

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Use local node as first priority to get structure of remote table when executing `cluster` and similar table functions.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Review by tavplubix — APPROVED** [review]
It's recommended to create `Distributed` table if you have many queries with specific cluster and remote tables. Tables functions like `cluster` are mostly for ad-hoc queries. But the improvement still makes sense, thanks!

**Comment by tavplubix** [comment]
@blinkov, something's wrong with CLAassistant bot

File: src/Storages/getStructureOfRemoteTable.cpp [hunk 1]
```diff
@@ -122,6 +122,19 @@ ColumnsDescription getStructureOfRemoteTable(
...
 
     std::string fail_messages;
+    
+    // use local shard as first priority, as it needs no network communication
+    for (const auto & shard_info : shards_info)
+    {
+        if(shard_info.isLocal())
+        {
+            const auto & res = getStructureOfRemoteTableInShard(cluster, shard_info, table_id, context, table_func_ptr);
+            if (res.empty())
```
Comment (azat) [inline src/Storages/getStructureOfRemoteTable.cpp:132]: This is not required for local nodes, it cannot be unavailable.

File: src/Storages/getStructureOfRemoteTable.cpp [hunk 2]
```diff
@@ -122,6 +122,19 @@ ColumnsDescription getStructureOfRemoteTable(
     const auto & shards_info = cluster.getShardsInfo();
 
     std::string fail_messages;
+    
+    // use local shard as first priority, as it needs no network communication
+    for (const auto & shard_info : shards_info)
+    {
+        if(shard_info.isLocal())
```
Comment (azat) [inline src/Storages/getStructureOfRemoteTable.cpp:129]:
```suggestion
        if (shard_info.isLocal())
```

File: src/Storages/getStructureOfRemoteTable.cpp (hunk 1 above)
Comment (azat) [reply src/Storages/getStructureOfRemoteTable.cpp:132]: Also I guess it is pretty easy to write a test, by looking at `query_log`

**Pull request merged**